### PR TITLE
Make auto mTLS works with workload level peer authentication policy.

### LINKED ISSUE #32788: auto mtls: add support for workload selectors on PeerAuthentication
Currently, auto mtls only works for namespace or cluster scoped PeerAuthentications. This makes creating a DISABLE PeerAuthentication for a specific workload really hard, since now you need a DR which is confusing.

cc @incfly

### PR DESCRIPTION
**Pull request by incfly**
Fixes #32788  #33809

- Creates the mTLS checker in the endpoint builder when the feature enabled.
- When policy enables mTLS plus the workload has the sidecar, we add metadata tlsMode: istio.
- Feature guarded via an env var. Integration test covered.

credits to @cpakulski for the work on envoy side

### CODE REVIEW

File: pilot/pkg/xds/endpoint_builder.go
Comment (yangminzhu) [inline pilot/pkg/xds/endpoint_builder.go:331]: update the log?

File: pilot/pkg/xds/endpoint_builder.go
Comment (yangminzhu) [inline pilot/pkg/xds/endpoint_builder.go:333]: just curious, what's the difference between `""` and `model.DisabledTLSModeLabel` for the `tlsMode`?

File: tests/integration/security/testdata/beta-mtls-automtls-workload.yaml
Comment (yangminzhu) [inline tests/integration/security/testdata/beta-mtls-automtls-workload.yaml:40]: 8090 or 8080?

File: tests/integration/security/reachability_test.go
Comment (yangminzhu) [inline tests/integration/security/reachability_test.go:109]: should the `ExpectMTLS` be used for the mTLS cases?

**Review by yangminzhu — COMMENTED** [review]
should also include a release note.

File: pilot/pkg/features/pilot.go
Comment (hzxuzhonghu) [inline pilot/pkg/features/pilot.go:150]: shouldn't it always be true?

File: pkg/test/framework/components/echo/annotations.go
Comment (howardjohn) [inline pkg/test/framework/components/echo/annotations.go:50]: what is this for?

File: tests/integration/security/testdata/beta-mtls-automtls-workload.yaml
Comment (incfly) [reply tests/integration/security/testdata/beta-mtls-automtls-workload.yaml:40]: 8090 is the http port, would just works fine, right?
https://github.com/istio/istio/blob/08f3ac1e1b7b8204dcc8fba2c0e0da5f6039536a/tests/integration/security/util/framework.go#L82

File: tests/integration/security/reachability_test.go
Comment (incfly) [reply tests/integration/security/reachability_test.go]: done. use mtlsOnExpect.

File: pkg/test/framework/components/echo/annotations.go
Comment (incfly) [reply pkg/test/framework/components/echo/annotations.go]: removed. i originally want to add support for per workload config option,but later find that might be not necessary and cause some trouble instead.

File: pilot/pkg/xds/endpoint_builder.go
Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go]: done.

File: pilot/pkg/xds/endpoint_builder.go
Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go:330]: it seems like only used in workload entry and service entry. not much difference in terms of the end effect AFAIK.

File: pilot/pkg/features/pilot.go
Comment (incfly) [reply pilot/pkg/features/pilot.go]: done.

File: tests/integration/security/testdata/beta-mtls-automtls-workload.yaml
Comment (yangminzhu) [reply tests/integration/security/testdata/beta-mtls-automtls-workload.yaml:40]: oh I thought this is commenting for the port number 8080 in line 41, maybe update the comment to make it clear

Comment (incfly) [reply tests/integration/security/testdata/beta-mtls-automtls-workload.yaml]: good point. updated to make the same.

File: pilot/pkg/xds/endpoint_builder.go
Comment (howardjohn) [inline pilot/pkg/xds/endpoint_builder.go:330]: should check the impact of this change (ie running benchmark, which I triggered)

**Review by howardjohn — COMMENTED** [review]
LGTM

File: pilot/pkg/xds/endpoint_builder.go
Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go:330]: sure. anything else you would suggest besides the benchmark results?

File: pilot/pkg/xds/endpoint_builder.go
Comment (stevenctl) [inline pilot/pkg/xds/endpoint_builder.go:328]: Does this also make sense to do with the client settings from DR?

**Review by stevenctl — APPROVED** [review]
lgtm! added hold pending benchmark results or other reviewers

File: pilot/pkg/xds/endpoint_builder.go
Comment (howardjohn) [reply pilot/pkg/xds/endpoint_builder.go:328]: If we do have client DR we shouldn't use auto mtls so it probably doesn't matter?

Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go:328]: you mean when the client side DR disable the mTLS? I think in that case the cds.go code would already configure the cluster as plaintext, no `transport_socket_matches` at all.

File: pilot/pkg/xds/endpoint_builder.go
Comment (howardjohn) [reply pilot/pkg/xds/endpoint_builder.go:330]:
```
EndpointGeneration/1/100-8                           376kB ± 0%         536kB ± 0%  +42.52%  (p=0.016 n=5+4)
EndpointGeneration/10/10-8                           246kB ± 0%         364kB ± 0%  +48.44%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                         2.33MB ± 0%        3.47MB ± 0%  +48.80%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-8                         2.34MB ± 0%        3.47MB ± 0%  +48.53%  (p=0.008 n=5+5)
```

Not great... do you think its feasible to rebuild only when tlsMode changes (which is very rare)?

**Comment by incfly** [comment]
The benchmark result is now available. Here are some cpu and mem delta for the EndpointGeneration. Noticed since there's no policy change in the test code  AFAIS, https://github.com/istio/istio/blob/39e6e3922878f292e89c8267aaa3c64cff3f6d38/pilot/pkg/xds/bench_test.go#L299. So these delta would only be the baseline delta cost for just checking the server endpoint mtls status, I think. The potential optimization, to only update metadata when the policy changed, would not change the numbers here.

not sure the criteria for such change and data. WDTY? @howardjohn @stevenctl 

```
EndpointGeneration/1000/1-8                         2.34MB ± 0%        3.47MB ± 0%  +48.53%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-8                         10.2ms ± 2%        12.4ms ± 3%  +21.15%  (p=0.008 n=5+5)

```

File: pilot/pkg/xds/endpoint_builder.go
Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go:330]: We definitely can do that(only rebuild when tls Mode changes), but that might not be that rare since now, `tlsMode` value can change caused by a peer authn config changed, right? See my previous comments.

Current code is building policy checker for each endpoint, which underneaths creates the PolicyApplier. I think reuse a single one for a single PushContext would reduce the memory significantly. From config perpective i think this would work as well since the policy applier is not endpoint specific.

Comment (howardjohn) [reply pilot/pkg/xds/endpoint_builder.go:330]: I think it is rare - we push endpoints FAR more often then peer auth changes. Likely many many orders of magnitude more often.

For PolicyApplier - doesn't it take `ep.Labels` into account?

Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go:330]: > I think it is rare in relationship to how often we hit this code path. We push endpoints FAR more often then peer auth changes. Likely many many orders of magnitude more often.

I see. you mean like when a single pod comes up we hit this code(rebuild the entire endpoint list for a given cluster maybe) which is far more often than the peer authn changes. In that case, if we only add these cost(cpu/memory) when peer authn changes, that could be acceptable.

true, the applier indeed take the labels into account, though we can build another map from labels to the applier instance to de-dup among the same set of endpoints.

Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go]: did some profilng againt the benchmark test. it appears that [1] the delta inc for the memory indeed comes the `buildEnvoyEndpoint`. conditionally construct the endpoint solves the problem. benchmark numbers should go back to the same level before.

image

File: pilot/pkg/xds/endpoint_builder.go
Comment (hzxuzhonghu) [inline pilot/pkg/xds/endpoint_builder.go:279]: why not construct when tlsMode != epTLSMode ?

File: pilot/pkg/xds/endpoint_builder.go
Comment (hzxuzhonghu) [reply pilot/pkg/xds/endpoint_builder.go:341]: I think we need to unify "" and DisabledTLSModeLabel, maybe later.

**Review by hzxuzhonghu — APPROVED** [review]

File: pilot/pkg/xds/endpoint_builder.go
Comment (incfly) [reply pilot/pkg/xds/endpoint_builder.go]: now fixed. see benchmark test results.

above comment is not enough. we still need to ensure when tlsMode == "", do not set filter metadata, just leave empty values. Because later on EndpointByNetworkFilters would use `proto.Clone`, increase the consumption as well.

**Comment by incfly** [comment]
Can you take another look? The benchmark data now looks okay to me. https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/34118/benchmark_istio/1417780463652573184/build-log.txt

**Review by howardjohn — APPROVED** [review]

**Comment by stevenctl** [comment]
@incfly does this need to be in 1.11?

**Comment by howardjohn** [comment]
Do not merge to 1.11: see https://github.com/istio/istio/issues/34227

**Pull request merged**
